Improve handling of dart format stderr output

natebosch opened the pull request:
Previously the `stdout` and `stderr` from the format execution were
merged and there was special handling for 1 specific warning message
followed by a standard 2 lines of boilerplate.

Now `stdout` an `stderr` are read separately. `stderr` is ignored unless
there is no output on `stdout`. This filters any warnings that didn't
prevent formatting, including the old hardcoded warning as well as a new
one around resolving `analysis_options.yaml` and any future warnings
that may come up.

When there is no `stdout` content check the `stderr` for the diagnostic
format. If any lines match, ignore any non-matching lines. If no lines
match, show the output anyway. I'm not currently aware of any output
here that we _wouldn't_ want to hide, but if error output becomes
important for the user to see we may need to revisit this design.

- Migrate from `systemlist` to `job_start` which allows separating the
  output streams.
- Pull the result handling out into a separate function so it can be the
  callback for when the output has been read.

natebosch committed:
Improve handling of dart format stderr output

Previously the `stdout` and `stderr` from the format execution were
merged and there was special handling for 1 specific warning message
followed by a standard 2 lines of boilerplate.

Now `stdout` an `stderr` are read separately. `stderr` is ignored unless
there is no output on `stdout`. This filters any warnings that didn't
prevent formatting, including the old hardcoded warning as well as a new
one around resolving `analysis_options.yaml` and any future warnings
that may come up.

When there is no `stdout` content treat the `stderr` as using the
diagnostic format. Ignore any lines that don't match the error format.
I'm not currently aware of any output here that we _wouldn't_ want to
hide, but if error output becomes important for the user to see we may
need to revisit this design.

- Migrate from `systemlist` to `job_start` which allows separating the
  output streams.
- Pull the result handling out into a separate function so it can be the
  callback for when the output has been read.

sigmundch commented:
I like your approach of ignoring stderr if there is a valid output in stdout. Do we have a guarantee that the output will be empty if any issue arises? Or do we need a secondary signal, like `dartfmt`'s exit code?

For the error modality, I worry we may hide an actionable warning/error, but I don't want to spam users with useless messages. Rather than always showing potentially non-actionable warning messages that don't match the diagnostic template, what if we only show extra warnings if we couldn't match any stderr line with the diagnostic template format? That would ensure that if there is no stdout, we always have something to share with users:

* non-empty stdout ==>  don't show any issues
* empty stdout + some stderr line matches diagnostic template ==> show diagnostic
* empty stdout + no stderr line matches diagnostic template ==> show the full stderr

natebosch commented:
> If it's annoying, we could add a flag to suppress it.

It's not actionable so I think its confusing to print it as a "Warning". Is there any reason not to suppress it always? Or to rewrite it as something like "Note: dart format does not read analysis file includes".

natebosch commented:
> what if we only show extra warnings if we couldn't match any stderr line with the diagnostic template format? That would ensure that if there is no stdout, we always have something to share with users:

This sounds like a good approach to me, I'll try it out.

mattrbeck committed:
Update installation instructions to suggest Vim 8 packages (#149)

This mirrors instructions for other popular plugins, like:
- https://github.com/tpope/vim-sensible
- https://github.com/leafgarland/typescript-vim

Move remaining config snippets to the wiki.

natebosch committed:
Use stdout as diagnostic if no lines match

natebosch committed:
Merge branch 'master' into dartfmt-stderr
